coreneuron/io/phase2.cpp

```diff
@@ -0,0 +1,1216 @@
+    else
+        nt._idata = nullptr;
+    // see patternstim.cpp
+    int extra_nv = (&nt == nrn_threads) ? nrn_extra_thread0_vdata : 0;
+    nt._nvdata = n_vdata;
+    if (nt._nvdata + extra_nv)
+        nt._vdata = (void**)ecalloc_align(nt._nvdata + extra_nv, sizeof(void*));
+    else
+        nt._vdata = nullptr;
+    // printf("_nidata=%d _nvdata=%d\n", nt._nidata, nt._nvdata);
+
+    // The data format begins with the matrix data
+    int ne = nrn_soa_padded_size(nt.end, MATRIX_LAYOUT);
+    size_t offset = 6 * ne;
+
+    if (n_diam) {
+        // in the rare case that a mechanism has dparam with diam semantics
+        // then actual_diam array added after matrix in nt._data
+        // Generally wasteful since only a few diam are pointed to.
+        // Probably better to move the diam semantics to the p array of the mechanism
+        offset += ne;
+    }
+
+    // Memb_list.data points into the nt.data array.
+    // Also count the number of Point_process
+    int num_point_process = 0;
+    for (auto tml = nt.tml; tml; tml = tml->next) {
+        Memb_list* ml = tml->ml;
+        int type = tml->index;
+        int layout = corenrn.get_mech_data_layout()[type];
+        int n = ml->nodecount;
+        int sz = nrn_prop_param_size_[type];
+        offset = nrn_soa_byte_align(offset);
+        ml->data = (double*)0 + offset;  // adjust below since nt._data not allocated
+        offset += nrn_soa_padded_size(n, layout) * sz;
+        if (corenrn.get_pnt_map()[type] > 0) {
+            num_point_process += n;
+        }
+    }
+    nt.pntprocs = (Point_process*)ecalloc_align(
+        num_point_process, sizeof(Point_process));  // includes acell with and without gid
+    nt.n_pntproc = num_point_process;
+    // printf("offset=%ld\n", offset);
+    nt._ndata = offset;
+
+    // now that we know the effect of padding, we can allocate data space,
+    // fill matrix, and adjust Memb_list data pointers
+    nt._data = (double*)ecalloc_align(nt._ndata, sizeof(double));
+    nt._actual_rhs = nt._data + 0 * ne;
+    nt._actual_d = nt._data + 1 * ne;
+    nt._actual_a = nt._data + 2 * ne;
+    nt._actual_b = nt._data + 3 * ne;
+    nt._actual_v = nt._data + 4 * ne;
+    nt._actual_area = nt._data + 5 * ne;
+    nt._actual_diam = n_diam ? nt._data + 6 * ne : nullptr;
+    for (auto tml = nt.tml; tml; tml = tml->next) {
+        Memb_list* ml = tml->ml;
+        ml->data = nt._data + (ml->data - (double*)0);
+    }
+
+    // matrix info
+    nt._v_parent_index = (int*)ecalloc_align(nt.end, sizeof(int));
+    std::copy(v_parent_index.begin(), v_parent_index.end(), nt._v_parent_index);
+    std::copy(actual_a.begin(), actual_a.end(), nt._actual_a);
+    std::copy(actual_b.begin(), actual_b.end(), nt._actual_b);
+    std::copy(actual_area.begin(), actual_area.end(), nt._actual_area);
+    std::copy(actual_v.begin(), actual_v.end(), nt._actual_v);
+    if (n_diam) {
+        std::copy(actual_diam.begin(), actual_diam.end(), nt._actual_diam);
+    }
```

Comment:
See comment in the .hpp file. We must avoid the doubly copy. It won't fit in the memory.